plpgsql: prevent a crash in an edge case during formatting

yuzefovich opened the pull request:
**tree: add panic-catching to formatting code**

We've seen a couple of node crashes due to runtime errors during
formatting AST. This behavior doesn't seem warranted, so this commit
adds the panic-catching mechanism to the main entry point of the
formatting code. We don't have an easy way to propagate the caught error
(this would require modifying almost a thousand of callsites), so we
simply write the error message into the buffer. Such behavior is
somewhat unfortunate, yet it seems better than a node crash. No panics
will be caught in the test builds.

Additionally, this commit improves our testing coverage somewhat so that
in test builds we now format each executed query unconditionally.

**plpgsql: prevent a crash in an edge case during formatting**

This commit fixes an issue that is similar to the one that was fixed in
22dabb08e76decbe2fc1c091431e05ebfba5e73f. In particular, each parser
(both SQL one and PLpgSQL one) keeps track of the number of annotations
found in the current stmt. Annotations are used during object name
resolution, and the AST nodes can reference annotation indices. If an
annotation index is set, then it assumes that during formatting the
corresponding annotation will be available in the supplied Annotations
container.

In the PLpgSQL parser we use an ephemeral SQL parser instance to process
SQL statements. Previously, we would simply discard the information about
whether any annotations were found in the SQL stmt which could lead to
undefined behavior down the line. This commit audits all usages of the
SQL parser from PLpgSQL one to ensure that we capture the number of
annotations used in that stmt and create the Annotations container with
that capacity, stored in the AST. This container is then substituted
into the FmtCtx for those ASTs. Note that since we currently don't
support DDLs from PLpgSQL, the annotation slots in the container won't
ever be set, so the contribution of this patch is preventing the crash.

Fixes: #143974.

Release note (bug fix): Previously, CockroachDB node could crash when
executing DO stmts when they contain currently unsupported DDL stmts like
CREATE TYPE in non-default configuration (additional logging like the one
controlled via `sql.log.all_statements.enabled` cluster setting needed
to be enabled). This bug was introduced in 25.1 release and is now
fixed.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/146406)
<!-- Reviewable:end -->

yuzefovich committed:
plpgsql: prevent a crash in an edge case during formatting

This commit fixes an issue that is similar to the one that was fixed in
22dabb08e76decbe2fc1c091431e05ebfba5e73f. In particular, each parser
(both SQL one and PLpgSQL one) keeps track of the number of annotations
found in the current stmt. Annotations are used during object name
resolution, and the AST nodes can reference annotation indices. If an
annotation index is set, then it assumes that during formatting the
corresponding annotation will be available in the supplied Annotations
container.

In the PLpgSQL parser we use an ephemeral SQL parser instance to process
SQL statements. Previously, we would simply discard the information about
whether any annotations were found in the SQL stmt which could lead to
undefined behavior down the line. This commit audits all usages of the
SQL parser from PLpgSQL one to ensure that we capture the number of
annotations used in that stmt and create the Annotations container with
that capacity, stored in the AST. This container is then substituted
into the FmtCtx for those ASTs. Note that since we currently don't
support DDLs from PLpgSQL, the annotation slots in the container won't
ever be set, so the contribution of this patch is preventing the crash.

Release note (bug fix): Previously, CockroachDB node could crash when
executing DO stmts when they contain currently unsupported DDL stmts like
CREATE TYPE in non-default configuration (additional logging like the one
controlled via `sql.log.all_statements.enabled` cluster setting needed
to be enabled). This bug was introduced in 25.1 release and is now
fixed.

yuzefovich committed:
tree: add panic-catching to formatting code

We've seen a couple of node crashes due to runtime errors during
formatting AST. This behavior doesn't seem warranted, so this commit
adds the panic-catching mechanism to the main entry point of the
formatting code. We don't have an easy way to propagate the caught error
(this would require modifying almost a thousand of callsites), so we
simply write the error message into the buffer. Such behavior is
somewhat unfortunate, yet it seems better than a node crash. No panics
will be caught in the test builds.

Additionally, this commit improves our testing coverage somewhat so that
in test builds we now format each executed query unconditionally.

Release note: None

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Thanks for the fix!

> Reviewed 3 of 3 files at r1, 7 of 7 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146406#-:-OPrD3JMBDRU4zyutzA5:b-72sm86)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @yuzefovich)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
TFTR!

bors r+
